File: zig/src/noise/session.zig
//! Session management for transport phase.
//!
//! NOT thread-safe. Caller must provide external synchronization
//! and pass timestamps explicitly. No std.Thread or std.time dependencies.

const std = @import("std");
const mem = std.mem;

const keypair = @import("keypair.zig");
const replay_mod = @import("replay.zig");
const crypto_mod = @import("crypto.zig");

pub const Key = keypair.Key;
pub const key_size = keypair.key_size;
pub const ReplayFilter = replay_mod.ReplayFilter;
pub const tag_size = crypto_mod.tag_size;

/// Session state.
pub const SessionState = enum {
    handshaking,
    established,
    expired,
};

/// Session timeout in nanoseconds (180 seconds).
pub const session_timeout_ns: u64 = 180 * std.time.ns_per_s;

/// Maximum nonce value.
pub const max_nonce: u64 = std.math.maxInt(u64) - 1;

/// Session errors.
pub const SessionError = error{
    NotEstablished,
    ReplayDetected,
    NonceExhausted,
    EncryptFailed,
    DecryptFailed,
    AuthenticationFailed,
};

/// Configuration for creating a session.
pub const SessionConfig = struct {
    local_index: u32,
    remote_index: u32 = 0,
    send_key: Key,
    recv_key: Key,
    remote_pk: Key = Key.zero,
    /// Current timestamp in nanoseconds (caller-provided).
    now_ns: u64 = 0,
};

/// Instantiate session for a given Crypto implementation and cipher suite.
pub fn SessionMod(comptime Crypto: type, comptime suite: crypto_mod.CipherSuite) type {
    const cipher = @import("cipher.zig").Cipher(Crypto, suite);

    return struct {
        /// An established Noise session with a peer.
        ///
        /// NOT thread-safe. Caller must synchronize access and provide
        /// timestamps via method parameters.
        pub const Session = struct {
            local_index: u32,
            remote_index: u32,

            send_key: Key,
            recv_key: Key,

            send_nonce: u64 = 0,
            recv_filter: ReplayFilter = ReplayFilter.init(),

            state: SessionState = .established,
            remote_pk: Key,

            created_ns: u64,
            last_received_ns: u64 = 0,
            last_sent_ns: u64 = 0,

            /// Creates a new session.
            pub fn init(cfg: SessionConfig) Session {
                return .{
                    .local_index = cfg.local_index,
                    .remote_index = cfg.remote_index,
                    .send_key = cfg.send_key,
                    .recv_key = cfg.recv_key,
                    .remote_pk = cfg.remote_pk,
                    .created_ns = cfg.now_ns,
                    .last_received_ns = cfg.now_ns,
                    .last_sent_ns = cfg.now_ns,
                };
            }

            pub fn localIndex(self: *const Session) u32 {
                return self.local_index;
            }

            pub fn remoteIndex(self: *const Session) u32 {
                return self.remote_index;
            }

            pub fn setRemoteIndex(self: *Session, idx: u32) void {
                self.remote_index = idx;
            }

            pub fn remotePk(self: *const Session) Key {
                return self.remote_pk;
            }

            pub fn getState(self: *const Session) SessionState {
                return self.state;
            }

            pub fn setState(self: *Session, new_state: SessionState) void {
                self.state = new_state;
            }

            /// Encrypts a message. Returns the nonce used.
            /// `now_ns` is the current timestamp in nanoseconds (caller-provided).
            pub fn encrypt(self: *Session, plaintext: []const u8, out: []u8, now_ns: u64) SessionError!u64 {
                if (self.state != .established) {
                    return SessionError.NotEstablished;
                }

                const nonce = self.send_nonce;
                if (nonce >= max_nonce) {
                    return SessionError.NonceExhausted;
                }
                self.send_nonce += 1;

                cipher.encrypt(&self.send_key.data, nonce, plaintext, "", out);
                self.last_sent_ns = now_ns;

                return nonce;
            }

            /// Decrypts a message.
            /// `now_ns` is the current timestamp in nanoseconds (caller-provided).
            pub fn decrypt(self: *Session, ciphertext: []const u8, nonce: u64, out: []u8, now_ns: u64) SessionError!usize {
                if (self.state != .established) {
                    return SessionError.NotEstablished;
                }

                if (!self.recv_filter.checkAndUpdate(nonce)) {
                    return SessionError.ReplayDetected;
                }

                if (ciphertext.len < tag_size) {
                    return SessionError.DecryptFailed;
                }

                cipher.decrypt(&self.recv_key.data, nonce, ciphertext, "", out) catch {
                    return SessionError.AuthenticationFailed;
                };

                self.last_received_ns = now_ns;

                return ciphertext.len - tag_size;
            }

            pub fn isExpired(self: *const Session, now_ns: u64) bool {
                if (self.state == .expired) return true;
                if (now_ns < self.last_received_ns) return false;
                return (now_ns - self.last_received_ns) > session_timeout_ns;
            }

            pub fn expire(self: *Session) void {
                self.state = .expired;
            }

            pub fn sendNonce(self: *const Session) u64 {
                return self.send_nonce;
            }

            pub fn recvMaxNonce(self: *const Session) u64 {
                return self.recv_filter.maxNonce();
            }

            pub fn createdNs(self: *const Session) u64 {
                return self.created_ns;
            }

            pub fn lastReceivedNs(self: *const Session) u64 {
                return self.last_received_ns;
            }

            pub fn lastSentNs(self: *const Session) u64 {
                return self.last_sent_ns;
            }
        };
    };
}

/// Creates a session index from 4 random bytes.
/// The caller is responsible for providing cryptographic random bytes.
pub fn generateIndexFromBytes(random_bytes: [4]u8) u32 {
    return mem.readInt(u32, &random_bytes, .little);
}

// Tests
const testing = std.testing;
const TestCrypto = @import("test_crypto.zig");
const TestSession = SessionMod(TestCrypto, .ChaChaPoly_BLAKE2s).Session;
const c = crypto_mod.CryptoMod(TestCrypto, .ChaChaPoly_BLAKE2s);

fn createTestSessions() struct { alice: TestSession, bob: TestSession } {
    const send_key = Key.fromBytes(c.hash(&.{"send key"}));
    const recv_key = Key.fromBytes(c.hash(&.{"recv key"}));

    const alice = TestSession.init(.{
        .local_index = 1,
        .remote_index = 2,
        .send_key = send_key,
        .recv_key = recv_key,
    });

    const bob = TestSession.init(.{
        .local_index = 2,
        .remote_index = 1,
        .send_key = recv_key,
        .recv_key = send_key,
    });

    return .{ .alice = alice, .bob = bob };
}

test "encrypt decrypt" {
    var sessions = createTestSessions();

    const plaintext = "Hello, World!";
    var ciphertext: [plaintext.len + tag_size]u8 = undefined;
    var decrypted: [plaintext.len]u8 = undefined;

    const nonce = try sessions.alice.encrypt(plaintext, &ciphertext, 0);
    const pt_len = try sessions.bob.decrypt(&ciphertext, nonce, &decrypted, 0);

    try testing.expectEqualSlices(u8, plaintext, decrypted[0..pt_len]);
}

test "nonce increment" {
    var sessions = createTestSessions();
    var ct: [4 + tag_size]u8 = undefined;

    for (0..10) |i| {
        try testing.expectEqual(@as(u64, i), sessions.alice.sendNonce());
        _ = try sessions.alice.encrypt("test", &ct, 0);
    }
    try testing.expectEqual(@as(u64, 10), sessions.alice.sendNonce());
}

test "replay protection" {
    var sessions = createTestSessions();

    const plaintext = "test";
    var ciphertext: [plaintext.len + tag_size]u8 = undefined;
    var decrypted: [plaintext.len]u8 = undefined;

    const nonce = try sessions.alice.encrypt(plaintext, &ciphertext, 0);
    _ = try sessions.bob.decrypt(&ciphertext, nonce, &decrypted, 0);
    try testing.expectError(SessionError.ReplayDetected, sessions.bob.decrypt(&ciphertext, nonce, &decrypted, 0));
}

test "state" {
    var sessions = createTestSessions();

    try testing.expectEqual(SessionState.established, sessions.alice.getState());
    sessions.alice.setState(.expired);
    try testing.expectEqual(SessionState.expired, sessions.alice.getState());

    var ct: [4 + tag_size]u8 = undefined;
    try testing.expectError(SessionError.NotEstablished, sessions.alice.encrypt("test", &ct, 0));
}

test "generate index from bytes" {
    const bytes1 = [4]u8{ 1, 0, 0, 0 };
    const bytes2 = [4]u8{ 2, 0, 0, 0 };
    try testing.expect(generateIndexFromBytes(bytes1) != generateIndexFromBytes(bytes2));
}

File: go/pkg/noise/session.go
package noise

import (
	"crypto/cipher"
	"crypto/rand"
	"encoding/binary"
	"errors"
	"sync"
	"sync/atomic"
	"time"

	"golang.org/x/crypto/chacha20poly1305"
)

// SessionState represents the current state of a session.
type SessionState uint32

const (
	// SessionStateHandshaking indicates the session is in the handshake phase.
	SessionStateHandshaking SessionState = iota
	// SessionStateEstablished indicates the session is ready for transport.
	SessionStateEstablished
	// SessionStateExpired indicates the session has expired and should not be used.
	SessionStateExpired
)

func (s SessionState) String() string {
	switch s {
	case SessionStateHandshaking:
		return "handshaking"
	case SessionStateEstablished:
		return "established"
	case SessionStateExpired:
		return "expired"
	default:
		return "unknown"
	}
}

const (
	// SessionTimeout is the duration after which a session expires without activity.
	SessionTimeout = 180 * time.Second

	// RekeyAfterMessages is the number of messages after which rekey should be triggered.
	RekeyAfterMessages = 1 << 60 // ~1 quintillion messages

	// MaxNonce is the maximum nonce value before the session must be rekeyed.
	MaxNonce = ^uint64(0) - 1 // Leave room to detect overflow
)

// Session represents an established Noise session with a peer.
// It manages the transport keys, nonces, and replay protection.
type Session struct {
	mu sync.RWMutex

	// Indices for message routing
	localIndex  uint32 // Our sender index
	remoteIndex uint32 // Peer's sender index

	// Transport keys and ciphers
	sendKey  Key
	recvKey  Key
	sendAEAD cipher.AEAD
	recvAEAD cipher.AEAD

	// Nonce management
	sendNonce atomic.Uint64
	recvNonce *ReplayFilter

	// State
	state    SessionState
	remotePK PublicKey // Remote peer's static public key

	// Timestamps
	createdAt    time.Time
	lastReceived time.Time
	lastSent     time.Time
}

// SessionConfig contains the configuration for creating a new session.
type SessionConfig struct {
	LocalIndex  uint32
	RemoteIndex uint32
	SendKey     Key
	RecvKey     Key
	RemotePK    PublicKey
}

// NewSession creates a new established session from handshake results.
func NewSession(cfg SessionConfig) (*Session, error) {
	sendAEAD, err := chacha20poly1305.New(cfg.SendKey[:])
	if err != nil {
		return nil, err
	}

	recvAEAD, err := chacha20poly1305.New(cfg.RecvKey[:])
	if err != nil {
		return nil, err
	}

	now := time.Now()
	s := &Session{
		localIndex:   cfg.LocalIndex,
		remoteIndex:  cfg.RemoteIndex,
		sendKey:      cfg.SendKey,
		recvKey:      cfg.RecvKey,
		sendAEAD:     sendAEAD,
		recvAEAD:     recvAEAD,
		recvNonce:    NewReplayFilter(),
		state:        SessionStateEstablished,
		remotePK:     cfg.RemotePK,
		createdAt:    now,
		lastReceived: now,
		lastSent:     now,
	}

	return s, nil
}

// LocalIndex returns the local sender index.
func (s *Session) LocalIndex() uint32 {
	return s.localIndex
}

// RemoteIndex returns the remote sender index.
func (s *Session) RemoteIndex() uint32 {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.remoteIndex
}

// SetRemoteIndex sets the remote sender index (during handshake).
func (s *Session) SetRemoteIndex(idx uint32) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.remoteIndex = idx
}

// RemotePublicKey returns the remote peer's static public key.
func (s *Session) RemotePublicKey() PublicKey {
	return s.remotePK
}

// State returns the current session state.
func (s *Session) State() SessionState {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.state
}

// SetState updates the session state.
func (s *Session) SetState(state SessionState) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state = state
}

// Encrypt encrypts a plaintext message for transport.
// Returns the ciphertext and the nonce used.
// The ciphertext includes the 16-byte authentication tag.
func (s *Session) Encrypt(plaintext []byte) (ciphertext []byte, nonce uint64, err error) {
	// Use read lock to get state and cipher (allows concurrent encryptions)
	s.mu.RLock()
	state := s.state
	sendAEAD := s.sendAEAD
	s.mu.RUnlock()

	if state != SessionStateEstablished {
		return nil, 0, ErrSessionNotEstablished
	}

	// Get and increment nonce atomically
	nonce = s.sendNonce.Add(1) - 1

	// Check for nonce exhaustion
	if nonce >= MaxNonce {
		return nil, 0, ErrNonceExhausted
	}

	// Build nonce bytes (12 bytes, little-endian)
	var nonceBytes [12]byte
	binary.LittleEndian.PutUint64(nonceBytes[:], nonce)

	// Encrypt (AEAD is safe for concurrent use)
	ciphertext = sendAEAD.Seal(nil, nonceBytes[:], plaintext, nil)

	// Update last sent time with write lock
	s.mu.Lock()
	s.lastSent = time.Now()
	s.mu.Unlock()

	return ciphertext, nonce, nil
}

// Decrypt decrypts a ciphertext message from transport.
// The nonce must be provided separately (from the packet header).
// Returns the plaintext if successful.
// Note: A corrupted packet that fails decryption will still consume its nonce
// in the replay filter. This is an acceptable trade-off for concurrent performance.
func (s *Session) Decrypt(ciphertext []byte, nonce uint64) ([]byte, error) {
	// Use read lock to get state and cipher (allows concurrent decryptions)
	s.mu.RLock()
	state := s.state
	recvAEAD := s.recvAEAD
	s.mu.RUnlock()

	if state != SessionStateEstablished {
		return nil, ErrSessionNotEstablished
	}

	// Atomically check and update replay filter
	if !s.recvNonce.CheckAndUpdate(nonce) {
		return nil, ErrReplayDetected
	}

	// Build nonce bytes
	var nonceBytes [12]byte
	binary.LittleEndian.PutUint64(nonceBytes[:], nonce)

	// Decrypt (AEAD is safe for concurrent use)
	plaintext, err := recvAEAD.Open(nil, nonceBytes[:], ciphertext, nil)
	if err != nil {
		return nil, ErrDecryptionFailed
	}

	// Update last received time with write lock
	s.mu.Lock()
	s.lastReceived = time.Now()
	s.mu.Unlock()

	return plaintext, nil
}

// IsExpired checks if the session has expired due to inactivity.
func (s *Session) IsExpired() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if s.state == SessionStateExpired {
		return true
	}

	return time.Since(s.lastReceived) > SessionTimeout
}

// Expire marks the session as expired.
func (s *Session) Expire() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.state = SessionStateExpired
}

// CreatedAt returns when the session was created.
func (s *Session) CreatedAt() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.createdAt
}

// LastReceived returns when the last message was received.
func (s *Session) LastReceived() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.lastReceived
}

// LastSent returns when the last message was sent.
func (s *Session) LastSent() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.lastSent
}

// SendNonce returns the current send nonce value.
func (s *Session) SendNonce() uint64 {
	return s.sendNonce.Load()
}

// RecvMaxNonce returns the highest received nonce.
func (s *Session) RecvMaxNonce() uint64 {
	return s.recvNonce.MaxNonce()
}

// Errors
var (
	ErrSessionNotEstablished = errors.New("noise: session not established")
	ErrReplayDetected        = errors.New("noise: replay detected")
	ErrNonceExhausted        = errors.New("noise: nonce exhausted, rekey required")
)

// GenerateIndex generates a random 32-bit index for session identification.
func GenerateIndex() (uint32, error) {
	var buf [4]byte
	if _, err := rand.Read(buf[:]); err != nil {
		return 0, err
	}
	return binary.LittleEndian.Uint32(buf[:]), nil
}
